crates/deltalake-core/src/kernel/snapshot/log_segment.rs

```diff
@@ -0,0 +1,665 @@
+use crate::operations::transaction::get_commit_bytes;
+use crate::protocol::DeltaOperation;
+use crate::{DeltaResult, DeltaTableConfig, DeltaTableError};
+
+use super::extract::{extract_and_cast, extract_and_cast_opt, read_primitive};
+use super::parse::{decode_stream, get_reader, read_from_json, read_metadata};
+
+const LAST_CHECKPOINT_FILE_NAME: &str = "_last_checkpoint";
+const BATCH_SIZE: usize = 1024;
+
+pub type CommitData = (Vec<Action>, DeltaOperation, Option<HashMap<String, Value>>);
+
+lazy_static! {
+    static ref CHECKPOINT_FILE_PATTERN: Regex =
+        Regex::new(r"\d+\.checkpoint(\.\d+\.\d+)?\.parquet").unwrap();
+    static ref DELTA_FILE_PATTERN: Regex = Regex::new(r"\d+\.json").unwrap();
+    pub(super) static ref COMMIT_SCHEMA: StructType = StructType::new(vec![
+        ActionType::Add.schema_field().clone(),
+        ActionType::Remove.schema_field().clone(),
+    ]);
+    pub(super) static ref CHECKPOINT_SCHEMA: StructType =
+        StructType::new(vec![ActionType::Add.schema_field().clone(),]);
+    pub(super) static ref TOMBSTONE_SCHEMA: StructType =
+        StructType::new(vec![ActionType::Remove.schema_field().clone(),]);
+}
+
+pub(super) trait PathExt {
+    fn child(&self, path: impl AsRef<str>) -> DeltaResult<Path>;
+    /// Returns the last path segment if not terminated with a "/"
+    fn filename(&self) -> Option<&str>;
+
+    /// Parse the version number assuming a commit json or checkpoint parquet file
+    fn commit_version(&self) -> Option<i64> {
+        self.filename()
+            .and_then(|f| f.split_once('.'))
+            .and_then(|(name, _)| name.parse().ok())
+    }
+
+    /// Returns true if the file is a checkpoint parquet file
+    fn is_checkpoint_file(&self) -> bool {
+        self.filename()
+            .map(|name| CHECKPOINT_FILE_PATTERN.captures(name).is_some())
+            .unwrap_or(false)
+    }
+
+    /// Returns true if the file is a commit json file
+    fn is_commit_file(&self) -> bool {
+        self.filename()
+            .map(|name| DELTA_FILE_PATTERN.captures(name).is_some())
+            .unwrap_or(false)
+    }
+}
+
+impl PathExt for Path {
+    fn child(&self, path: impl AsRef<str>) -> DeltaResult<Path> {
+        Ok(self.child(path.as_ref()))
+    }
+
+    fn filename(&self) -> Option<&str> {
+        self.filename()
+    }
+}
+
+#[inline]
+fn get_decoder(schema: ArrowSchemaRef, config: &DeltaTableConfig) -> DeltaResult<Decoder> {
+    Ok(ReaderBuilder::new(schema)
+        .with_batch_size(config.log_batch_size)
+        .build_decoder()?)
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct FileInfo {
+    location: String,
+    size: usize,
+    last_modified: chrono::DateTime<Utc>,
+    e_tag: Option<String>,
+    version: Option<i64>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(super) struct LogSegment {
+    pub(super) version: i64,
+    pub(super) commit_files: VecDeque<ObjectMeta>,
+    pub(super) checkpoint_files: Vec<ObjectMeta>,
+}
```

Comment:
I've been trying to wrap my head around this and the benefit obtained from storing the ObjectMeta for commit files.
Suppose we want to load a table at version 60. If there is no checkpoint then we can infer &check the json files from 0 to 60 exist and then in the LogSegment store this range.
If the table had a checkpoint at version 50 then we can store & check the files files from 51 to 60 and store that range. 
From what I've seen we can remove `commit_files` and replace it with `commit_json_start` which is the first json file that must be read after the snapshot. If the snapshot matches the version then it can be None.

Reply:
This gets a bit more complex, as we need to know which checkpoint files to read. I.e. while deprecated we should still be able to read multi-part checkpoints. With V2 checkpoints (not yet supported) we also need to be able to read the sidecar files.

More broadly speaking, for parquet files we always require the file size to read it, so we avoid an extra `head` if we store the `ObjectMeta` from list. 

I do see the point though that we may want to see if we can avoid storing some of these files in the final snapshot.

Reply:
Wrinting some comments I realized we also do this to support time travel right now. There we rely on the modification time of the log files to identify the right version number.